### DISCUSSION TITLE
Voyager Story in web context

### DISCUSSION THREAD

**Discussion by dirkroorda** [Smithsonian/dpo-voyager#146] [General]
Not sure whether an issue is the right medium to ask a question like the following:

We (Pure3D project) like to use Voyager Story in a web context.
The scenario is: 

1) a user visits a page where (s)he can select a 3d model and optionally a set of annotations
2) the webserver serves Voyager Story (as single page app) with the model loaded and the annotations loaded
3) the users interacts with the viewer, edits annotations, and saves his/her work
4) Voyager then sends the saved annotations back to the server as JSON
5) The server processes the annotations and stores/merges them in a database

Does the current Voyager code supports this interaction? The critical points are:

A) can we serve Voyager Story with model and annotations already selected, so that the user does not have to do that?
B) does the Voyager Story client submit the saved results to the server, and should we embed the voyager client in a <form> element?

As you see, we are trying to use the voyager client, but want to control the serving stuff ourselves.
Thanks in advance!

**Comment by gjcope** [comment]
Hi @dirkroorda , thanks for the questions. I went ahead and converted the issue to a discussion.

I will address your points below:

1. Voyager Story (and Explorer) have a 'root' and 'document' attribute that define the scene to be loaded. So it is no problem to have a page that allows you to select different scenes via a customized url. An example of this is the 'explore' page on our site: https://3d.si.edu/explore - However, annotations are contained within that scene file, so selecting a scene and annotations independent of each other would be functionality that does not currently exist, though you can have multiple sets of annotations in one scene file.
2. Yes, that is how it would function.
3. Yes
4. Yes, Voyager would send the full scene json file back to the server via a WebDAV PUT request which would include any changes made to the scene, annotations and all.
5. If the annotations are to be stored separately from the scene file, you would need to strip them out server side, or pull them out with our annotation api function and store them independently of Voyager.

A) Yes, as described above.
B) Yes, the saved scene file is submitted to the server. Yes, the Voyager Story client would be embedded as custom web component.

Please let me know if I can make any more clarifications!

**Reply by dirkroorda** [reply]
Thanks, Jamie!
As to 5: indeed we plan to do that, because multiple users will make annotations to the same model. In order to make that work, the server adds the creator and time of modification to each annotation, in order to merge them all properly later on.

**Reply by dirkroorda** [reply]
So, any time a user requests a model + scene, the scene will be distilled on the fly from the database and sent to the client.
We remember what we sent to which user and when.
Any time a full scene files comes back, we compare it with what has been sent to that user, en filter out the changes, and attribute the changes to that user.
Then we merge the changes into the database.
If we have to update something that has been changed by another user in the meantime, we flag a save conflict.
Editors can resolve save conflicts.
I think we can make this work, based on the information that the webserver collects from the traffic between client and server.

**Reply by dirkroorda** [reply]
Or am I redoing webdav now?

**Reply by dirkroorda** [reply]
I am slowly getting it: the voyager client acts as a webdav client that writes to a remote file system. So the webserver should offer a filesystem through the webdav protocol and will see the results coming in.
Then the question is: can we constrain the file locations that get written to by the voyager client?

And the other question is: how can the webserver detect that a webdav put has happened?

I wonder if there is also a mode in the voyager where the scene is not transmitted by means of a webdav put but by means of a submit or asynchronous request with a json payload?

**Reply by gjcope** [reply]
Yes, exactly. You would need the WebDAV enabled file server to respond to requests and do any custom action you need on the server side. As designed, Voyager will only write to the location assigned to 'root' where the project files are picked up from and I assume restrictions could be put in place on the server side as well. @jahjedtieson is on vacation, but implemented something like this for Packrat where urls are rerouted and custom logic happens. He may have some further insights on how best to proceed on that front.

Some functions of Voyager rely on the extended features of WebDAV. It's certainly possible to have a functional client without those features, but it would take some development.

**Reply by jahjedtieson** [reply]
Hi @dirkroorda and @gjcope.  Indeed, https://github.com/Smithsonian/dpo-packrat has implemented a WebDAV server for use in handling WebDAV puts coming from Voyager Story.  In handling these, we encountered some slightly tricky situations in which we had to ensure all of the inbound bits had been received before Packrat took action to process those bits (i.e. ingest them into the repository).  This issue was solely a result of the 3rd party components used by Packrat to implement its WebDAV server -- Voyager appears to be a very well-behaved WebDAV client.

I would be happy to dig into any specific questions you may have about Packrat's implementation here!

**Reply by dirkroorda** [reply]
@gjcope Exactly, rerouting of urls and custom logic will be exactly what we need if we serve voyager to annotating users.
Thanks @jahjedtieson I'll have a look at that. With a special look as to how we prevent one user overwriting the annotations of another user. Or, how to make that possible in those cases where we want exactly that.
